undefined behavior in argument evaluation for multiple-argument functions

Test code:
```
struct MyType
{
	void myFunc(int, int);
};

LuaClass<MyType> myLuaClass(l, "myClass");
myLuaClass.fun("myFunc", &MyType::myFunc);
```
Produces this warning:
```
In file included from ./controllers/lua/lua_hooks.cpp:12:
./controllers/lua/luaaa/luaaa.hpp: In instantiation of 'static int luaaa::MemberFunctionCaller(void (TCLASS::*)(ARGS ...))::HelperClass::Invoke(lua_State*) [with TCLASS = configureRusefiLuaHooks(lua_State*)::MyType; ARGS = {int, int}; lua_State = lua_State]':
./controllers/lua/luaaa/luaaa.hpp:499:3:   required from 'int (* luaaa::MemberFunctionCaller(void (TCLASS::*)(ARGS ...)))(lua_State*) [with TCLASS = configureRusefiLuaHooks(lua_State*)::MyType; ARGS = {int, int}; lua_CFunction = int (*)(lua_State*); lua_State = lua_State]'
./controllers/lua/luaaa/luaaa.hpp:641:50:   required from 'luaaa::LuaClass< <template-parameter-1-1> >& luaaa::LuaClass< <template-parameter-1-1> >::fun(const char*, F) [with F = void (configureRusefiLuaHooks(lua_State*)::MyType::*)(int, int); TCLASS = configureRusefiLuaHooks(lua_State*)::MyType]'
./controllers/lua/lua_hooks.cpp:431:42:   required from here
./controllers/lua/luaaa/luaaa.hpp:495:6: error: operation on 'idx' may be undefined [-Werror=sequence-point]
  495 |      (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(LuaStack<ARGS>::get(state, idx--)...);
      |      ^
```

As you can see, this line in `luaaa.hpp` is the issue:
`(LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(LuaStack<ARGS>::get(state, idx--)...);`
The list of arguments is evaluated one at a time via the parameter pack, then the arguments are passed to the underlying function.  The problem is that in C++, it is undefined behavior to assign the same variable more than once between any two sequence points.  The first sequence point is the previous line, and the next sequence point is the call to the function returned from `LuaStack::get`, inbetween which `idx` is assigned once for every argument passed.

PS: The fact that the commit 51ab2ef54dd4d8678e28d489805176a84a19a9cc has different behavior for GCC vs. Clang vs. MSVC indicates that you're relying on undefined behavior that may not even be consistent within one particular compiler.

--- a/luaaa.hpp
+++ b/luaaa.hpp
@@ -444,14 +444,14 @@ namespace LUAAA_NS
     }
 
 	template<typename FTYPE, typename ...ARGS>
-	void LuaInvokeVoid(lua_State* state, void* calleePtr, size_t skip)
+	inline void LuaInvokeVoid(lua_State* state, void* calleePtr, size_t skip)
 	{
 		int idx = 0; (void)(idx);
 		(*(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), skip, &idx)...);
 	}
 
 	template<typename TRET, typename FTYPE, typename ...ARGS>
-	TRET LuaInvoke(lua_State* state, void* calleePtr, size_t skip)
+	inline TRET LuaInvoke(lua_State* state, void* calleePtr, size_t skip)
 	{
 		int idx = 0; (void)(idx);
 		return (*(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), skip, &idx)...);
@@ -550,14 +550,14 @@ namespace LUAAA_NS
 	// member function invoker
 	//========================================================
 	template<typename TCLASS, typename TRET, typename FTYPE, typename ...ARGS>
-	TRET LuaInvokeInstanceMember(lua_State* state, void* calleePtr)
+	inline TRET LuaInvokeInstanceMember(lua_State* state, void* calleePtr)
 	{
 		int idx = 0; (void)(idx);
 		return (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...);
 	}
 
 	template<typename TCLASS, typename FTYPE, typename ...ARGS>
-	void LuaInvokeInstanceMemberVoid(lua_State* state, void* calleePtr)
+	inline void LuaInvokeInstanceMemberVoid(lua_State* state, void* calleePtr)
 	{
 		int idx = 0; (void)(idx);
 		(LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...);

--- a/luaaa.hpp
+++ b/luaaa.hpp
@@ -443,6 +443,20 @@ namespace LUAAA_NS
         return (*getter)(state, offset);
     }
 
+	template<typename FTYPE, typename ...ARGS>
+	void LuaInvokeVoid(lua_State* state, void* calleePtr, size_t skip)
+	{
+		int idx = 0; (void)(idx);
+		(*(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), skip, &idx)...);
+	}
+
+	template<typename TRET, typename FTYPE, typename ...ARGS>
+	TRET LuaInvoke(lua_State* state, void* calleePtr, size_t skip)
+	{
+		int idx = 0; (void)(idx);
+		return (*(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), skip, &idx)...);
+	}
+
 #define IMPLEMENT_FUNCTION_CALLER(CALLERNAME, CALLCONV, SKIPPARAM) \
     template<typename TRET, typename ...ARGS> \
 	lua_CFunction CALLERNAME(TRET(CALLCONV*func)(ARGS...)) \
@@ -456,8 +470,7 @@ namespace LUAAA_NS
 				luaL_argcheck(state, calleePtr, 1, "cpp closure function not found."); \
 				if (calleePtr) \
 				{ \
-					int idx = 0; (void)(idx); \
-					LuaStackReturn<TRET>(state, (*(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), SKIPPARAM, &idx)...)); \
+					LuaStackReturn<TRET>(state, LuaInvoke<TRET, FTYPE, ARGS...>(state, calleePtr, SKIPPARAM)); \
 					return 1; \
 				} \
 				return 0; \
@@ -477,8 +490,7 @@ namespace LUAAA_NS
 				luaL_argcheck(state, calleePtr, 1, "cpp closure function not found."); \
 				if (calleePtr) \
 				{ \
-					int idx = 0; (void)(idx); \
-					(*(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), SKIPPARAM, &idx)...); \
+					LuaInvokeVoid<FTYPE, ARGS...>(state, calleePtr, SKIPPARAM); \
 				} \
 				return 0; \
 			} \
@@ -537,6 +549,20 @@ namespace LUAAA_NS
 	//========================================================
 	// member function invoker
 	//========================================================
+	template<typename TCLASS, typename TRET, typename FTYPE, typename ...ARGS>
+	TRET LuaInvokeInstanceMember(lua_State* state, void* calleePtr)
+	{
+		int idx = 0; (void)(idx);
+		return (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...);
+	}
+
+	template<typename TCLASS, typename FTYPE, typename ...ARGS>
+	void LuaInvokeInstanceMemberVoid(lua_State* state, void* calleePtr)
+	{
+		int idx = 0; (void)(idx);
+		(LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...);
+	}
+
     template<typename TCLASS, typename TRET, typename ...ARGS>
     lua_CFunction MemberFunctionCaller(TRET(TCLASS::*func)(ARGS...))
     {
@@ -550,7 +576,7 @@ namespace LUAAA_NS
                 if (calleePtr)
                 {
                     int idx = 0; (void)(idx);
-                    LuaStackReturn<TRET>(state, (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...));
+                    LuaStackReturn<TRET>(state, LuaInvokeInstanceMember<TCLASS, TRET, FTYPE, ARGS...>(state, calleePtr));
                     return 1;
                 }
                 return 0;
@@ -572,7 +598,7 @@ namespace LUAAA_NS
                 if (calleePtr)
                 {
                     int idx = 0; (void)(idx);
-                    LuaStackReturn<TRET>(state, (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...));
+                    LuaStackReturn<TRET>(state, LuaInvokeInstanceMember<TCLASS, TRET, FTYPE, ARGS...>(state, calleePtr));
                     return 1;
                 }
                 return 0;
@@ -593,8 +619,7 @@ namespace LUAAA_NS
                 luaL_argcheck(state, calleePtr, 1, "cpp closure function not found.");
                 if (calleePtr)
                 {
-                    int idx = 0; (void)(idx);
-                    (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...);
+                    LuaInvokeInstanceMemberVoid<TCLASS, FTYPE, ARGS...>(state, calleePtr);
                 }
                 return 0;
             }
@@ -614,8 +639,7 @@ namespace LUAAA_NS
                 luaL_argcheck(state, calleePtr, 1, "cpp closure function not found.");
                 if (calleePtr)
                 {
-                    int idx = 0; (void)(idx);
-                    (LuaStack<TCLASS>::get(state, 1).**(FTYPE*)(calleePtr))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 1, &idx)...);
+                    LuaInvokeInstanceMemberVoid<TCLASS, FTYPE, ARGS...>(state, calleePtr);
                 }
                 return 0;
             }
@@ -775,8 +799,7 @@ namespace LUAAA_NS
                     void * spawner = lua_touserdata(state, lua_upvalueindex(1));
                     luaL_argcheck(state, spawner, 1, "cpp closure spawner not found.");
                     if (spawner) {
-                        int idx = 0; (void)(idx);
-                        auto obj = (*(SPAWNERFTYPE*)(spawner))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 0, &idx)...);
+                        auto obj = LuaInvoke<TCLASS*, SPAWNERFTYPE, ARGS...>(state, spawner, 0);
                         if (obj)
                         {
                             TCLASS ** objPtr = (TCLASS**)lua_newuserdata(state, sizeof(TCLASS*));
@@ -859,8 +882,7 @@ namespace LUAAA_NS
                     luaL_argcheck(state, deleter, 1, "cpp closure deleter not found.");
 
                     if (spawner) {
-                        int idx = 0; (void)(idx);
-                        auto obj = (*(SPAWNERFTYPE*)(spawner))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 0, &idx)...);
+                        auto obj = LuaInvoke<TCLASS*, SPAWNERFTYPE, ARGS...>(state, spawner, 0);
                         if (obj)
                         {
                             TCLASS ** objPtr = (TCLASS**)lua_newuserdata(state, sizeof(TCLASS*));
@@ -952,8 +974,7 @@ namespace LUAAA_NS
                     void * spawner = lua_touserdata(state, lua_upvalueindex(1));
                     luaL_argcheck(state, spawner, 1, "cpp closure spawner not found.");
                     if (spawner) {
-                        int idx = 0; (void)(idx);
-                        auto obj = (*(SPAWNERFTYPE*)(spawner))(stackOperatorCaller(state, LuaStack<ARGS>::get, sizeof...(ARGS), 0, &idx)...);
+                        auto obj = LuaInvoke<TCLASS*, SPAWNERFTYPE, ARGS...>(state, spawner, 0);
                         if (obj)
                         {
                             TCLASS ** objPtr = (TCLASS**)lua_newuserdata(state, sizeof(TCLASS*));

--- a/luaaa.hpp
+++ b/luaaa.hpp
@@ -575,7 +575,6 @@ namespace LUAAA_NS
                 luaL_argcheck(state, calleePtr, 1, "cpp closure function not found.");
                 if (calleePtr)
                 {
-                    int idx = 0; (void)(idx);
                     LuaStackReturn<TRET>(state, LuaInvokeInstanceMember<TCLASS, TRET, FTYPE, ARGS...>(state, calleePtr));
                     return 1;
                 }
@@ -597,7 +596,6 @@ namespace LUAAA_NS
                 luaL_argcheck(state, calleePtr, 1, "cpp closure function not found.");
                 if (calleePtr)
                 {
-                    int idx = 0; (void)(idx);
                     LuaStackReturn<TRET>(state, LuaInvokeInstanceMember<TCLASS, TRET, FTYPE, ARGS...>(state, calleePtr));
                     return 1;
                 }

